k2/csrc/tensor_ops.cu

```diff
@@ -507,4 +507,249 @@ Tensor SimpleRaggedIndexSelect1D(Tensor &src, Ragged<int32_t> &indexes) {
+    Real cur_sum = 0.0;
+    for (int32_t t = 0; t < T; t++) {
+      cur_sum = x[t * stride1] + cur_sum * gamma[t * stride1];
+      y[t * stride1] = cur_sum;
+    }
+  }
+}
+
+
+void DiscountedCumSum(const Tensor &src, const Tensor &gamma, Tensor *dest) {
+  // check contexts compatible:
+  if (!(IsCompatible(src, gamma) && IsCompatible(src, *dest))) {
+    K2_LOG(ERROR) << "Tensors are on different devices";
+  }
+  if (!(src.NumAxes() == 2 && gamma.NumAxes() == 2 && dest->NumAxes() == 2)) {
+    K2_LOG(ERROR) << "Expected all num-axes to equal 2.";
+  }
+  if (!(src.SameDims(gamma) && src.SameDims(*dest))) {
+    K2_LOG(ERROR) << "Expected all args to have the same dim.";
+  }
+  if (!(src.Stride(1) == gamma.Stride(1) && src.Stride(1) == dest->Stride(1))) {
+    K2_LOG(ERROR) << "Expected all strides on dim 1 to be the same.";
+  }
+  if (!(src.GetDtype() == gamma.GetDtype() && src.GetDtype() == dest->GetDtype())) {
+    K2_LOG(ERROR) << "Expected all args to have the same dtype.";
+  }
+  int32_t N = src.Dim(0),
+      T = src.Dim(1),
+      src_stride0 = src.Stride(0),
+      gamma_stride0 = gamma.Stride(0),
+      dest_stride0 = dest->Stride(0),
+      stride1 = src.Stride(1);  // these are all the same.
+  ContextPtr c = src.Context();
+  if (src.GetDtype() == kFloatDtype) {
+    if (c->GetDeviceType() == kCuda) {
+      DiscountedCumSumCudaImpl<float, 128>(c->GetCudaStream(), N, T,
+                                           src.Data<float>(), src_stride0,
+                                           gamma.Data<float>(), gamma_stride0,
+                                           dest->Data<float>(), dest_stride0,
+                                           stride1);
+    } else {
+      DiscountedCumSumCpuImpl<float>(N, T,
+                                     src.Data<float>(), src_stride0,
+                                     gamma.Data<float>(), gamma_stride0,
+                                     dest->Data<float>(), dest_stride0,
+                                     stride1);
+
+    }
+  } else if (src.GetDtype() == kDoubleDtype) {
+    if (c->GetDeviceType() == kCuda) {
+      DiscountedCumSumCudaImpl<double, 128>(c->GetCudaStream(), N, T,
+                                            src.Data<double>(), src_stride0,
+                                            gamma.Data<double>(), gamma_stride0,
+                                            dest->Data<double>(), dest_stride0,
+                                            stride1);
+    } else {
+      DiscountedCumSumCpuImpl<double>(N, T,
+                                      src.Data<double>(), src_stride0,
+                                      gamma.Data<double>(), gamma_stride0,
+                                      dest->Data<double>(), dest_stride0,
+                                      stride1);
+
+    }
+  } else {
+        K2_LOG(ERROR) << "This algorithm only instantiated for float and double; type is "
```

Comment:
I would suggest `K2_LOG(FATAL)` to terminate the program (i.e., to throw an exception).